From: Peter Kokot <peterkokot@gmail.com>
Subject: Add CMake changes for PHP-8.5 branch

- EditorConfig adjustments for code style support in editors and IDEs
- CMake build system files added to .gitignore
- CMake added to run-tests.php for info about failing tests and checks
- CMake-related modifications added to ext/skeleton template directory
  and ext/ext_skel.php script
---
 .editorconfig              |  2 +-
 .gitignore                 | 50 ++++++++++++++++++++++++++++++++++++++
 ext/ext_skel.php           |  2 ++
 ext/skeleton/.gitignore.in | 23 ++++++++++++++++++
 run-tests.php              | 24 +++++++++++++-----
 5 files changed, 94 insertions(+), 7 deletions(-)

diff --git a/.editorconfig b/.editorconfig
index 19be96087d..58a83dfb68 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -17,7 +17,7 @@ indent_style             = tab
 indent_size              = 4
 indent_style             = space
 
-[*.{ac,m4,sh,yml}]
+[{CMakeLists.{txt,txt.in},*.{ac,cmake,cmake.in,json,m4,sh,yml}}]
 indent_size              = 2
 indent_style             = space
 
diff --git a/.gitignore b/.gitignore
index 52dde98ed4..d698b5a19c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -152,6 +152,7 @@ php
 # ------------------------------------------------------------------------------
 /ext/json/json_parser.tab.h
 /ext/json/json_parser.tab.c
+/ext/json/json_parser.output
 /sapi/phpdbg/phpdbg_parser.c
 /sapi/phpdbg/phpdbg_parser.h
 /sapi/phpdbg/phpdbg_parser.output
@@ -290,9 +291,58 @@ tmp-php.ini
 /junit.out.xml
 /.ccache/
 
+# ------------------------------------------------------------------------------
+# CMake-based build system files
+# ------------------------------------------------------------------------------
+cmake_install.cmake
+CMakeCache.txt
+CMakeFiles/
+
+# Generated by FetchContent
+/_deps/
+
+# Graphviz generated files
+/*.dependers
+/*.dot
+
+# Generated by the Ninja build system
+/.ninja*
+/build.ninja
+
+# Local user presets
+/CMakeUserPresets.json
+
+# Generated when CMAKE_EXPORT_COMPILE_COMMANDS is enabled
+/compile-commands.json
+
+# Generated by ctest
+/CTestTestfile.cmake
+/Testing/
+
+# CMake script profiling data output (--profiling-output <path>)
+/profile.json
+
+# Generated by QT Creator
+CMakeLists.txt.user
+
+# Generated by XCode
+CMakeScripts/
+
+# Generated by cmake --install
+/install_manifest.txt
+
+# Generated by cmake when cross-compiling if missing cache variables are found
+/TryRunResults.cmake
+
+# pkg-config .pc files
+/sapi/embed/php-embed.pc
+/scripts/php.pc
+
 # ------------------------------------------------------------------------------
 # Special cases to invert previous ignore patterns
 # ------------------------------------------------------------------------------
+!**/cmake/config.h.in
+!**/cmake/modules/
 !/ext/bcmath/libbcmath/src/config.h
 !/ext/fileinfo/libmagic/config.h
 !/ext/fileinfo/libmagic.patch
diff --git a/ext/ext_skel.php b/ext/ext_skel.php
index cd82abb346..74ee94c7e2 100755
--- a/ext/ext_skel.php
+++ b/ext/ext_skel.php
@@ -289,6 +289,8 @@ function copy_config_scripts() {
         $files[] = 'config.w32';
     }
 
+    $files[] = 'CMakeLists.txt';
+    $files[] = 'cmake/config.h.in';
     $files[] = '.gitignore';
 
     foreach($files as $config_script) {
diff --git a/ext/skeleton/.gitignore.in b/ext/skeleton/.gitignore.in
index e691bd3964..191fcd70eb 100644
--- a/ext/skeleton/.gitignore.in
+++ b/ext/skeleton/.gitignore.in
@@ -45,3 +45,26 @@ tests/**/*.db
 tests/**/*.mem
 tmp-php.ini
 *~
+
+# ------------------------------------------------------------------------------
+# CMake-based build system files
+# ------------------------------------------------------------------------------
+!**/cmake/config.h.in
+!**/cmake/modules/
+/_deps/
+/.ninja*
+/*.dependers
+/*.dot
+/build.ninja
+/CMakeUserPresets.json
+/compile-commands.json
+/CTestTestfile.cmake
+/install_manifest.txt
+/profile.json
+/Testing/
+/TryRunResults.cmake
+cmake_install.cmake
+CMakeCache.txt
+CMakeFiles/
+CMakeLists.txt.user
+CMakeScripts/
diff --git a/run-tests.php b/run-tests.php
index 0ff11c2c95..f64002f185 100755
--- a/run-tests.php
+++ b/run-tests.php
@@ -847,10 +847,12 @@ function write_information(array $user_tests, $phpdbg): void
         <?php
         $exts = get_loaded_extensions();
         $ext_dir = ini_get('extension_dir');
-        foreach (scandir($ext_dir) as $file) {
-            if (preg_match('/^(?:php_)?([_a-zA-Z0-9]+)\.(?:so|dll)$/', $file, $matches)) {
-                if (!extension_loaded($matches[1])) {
-                    $exts[] = $matches[1];
+        if (is_dir($ext_dir)) {
+            foreach (scandir($ext_dir) as $file) {
+                if (preg_match('/^(?:php_)?([_a-zA-Z0-9]+)\.(?:so|dll)$/', $file, $matches)) {
+                    if (!extension_loaded($matches[1])) {
+                        $exts[] = $matches[1];
+                    }
                 }
             }
         }
@@ -939,7 +941,7 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
 
     $failed_tests_data .= "\n" . $sep . 'BUILD ENVIRONMENT' . $sep;
     $failed_tests_data .= "OS:\n" . PHP_OS . " - " . php_uname() . "\n\n";
-    $ldd = $autoconf = $sys_libtool = $libtool = $compiler = 'N/A';
+    $ldd = $autoconf = $sys_libtool = $libtool = $compiler = $cmake = 'N/A';
 
     if (!IS_WINDOWS) {
         /* If PHP_AUTOCONF is set, use it; otherwise, use 'autoconf'. */
@@ -950,7 +952,9 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
         }
 
         /* Always use the generated libtool - Mac OSX uses 'glibtool' */
-        $libtool = shell_exec(INIT_DIR . '/libtool --version');
+        if (file_exists(INIT_DIR . '/libtool')) {
+            $libtool = shell_exec(INIT_DIR . '/libtool --version');
+        }
 
         /* Use shtool to find out if there is glibtool present (MacOSX) */
         $sys_libtool_path = shell_exec(__DIR__ . '/build/shtool path glibtool libtool');
@@ -972,9 +976,17 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
         }
 
         $ldd = shell_exec("ldd $php 2>/dev/null");
+
+        /** If PHP_CMAKE is set, use it; otherwise, use 'cmake'. */
+        if (getenv('PHP_CMAKE')) {
+            $cmake = shell_exec(getenv('PHP_CMAKE') . ' --version');
+        } else {
+            $cmake = shell_exec('cmake --version');
+        }
     }
 
     $failed_tests_data .= "Autoconf:\n$autoconf\n";
+    $failed_tests_data .= "CMake:\n$cmake\n";
     $failed_tests_data .= "Bundled Libtool:\n$libtool\n";
     $failed_tests_data .= "System Libtool:\n$sys_libtool\n";
     $failed_tests_data .= "Compiler:\n$compiler\n";
